src/main/java/biz/paluch/logging/gelf/intern/sender/GelfTCPSender.java

```diff
@@ -138,10 +159,35 @@ protected void write(ByteBuffer buffer) throws IOException {
                 // indicator the socket was closed
                 Closer.close(channel());
                 throw new SocketException("Cannot write buffer to channel");
+            } if (written == 0) {
+                nothingWrittenTimesInRow++;
+                if (nothingWrittenTimesInRow > writeBackoffThreshold) {
+                    if (totalSleepTimeMs > maxWriteBackoffTimeMs) {
+                        Closer.close(channel());
+                        throw new SocketException("Cannot write buffer to channel, no progress in writing");
+                    }
+                    totalSleepTimeMs += sleep(writeBackoffTimeMs);
+                }
+            } else { // written > 0
+                nothingWrittenTimesInRow = 0;
             }
         }
     }
 
+    private static long sleep(long millis) {
+        long startTime = System.nanoTime();
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+        long sleepTimeMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime);
```

Comment:
Nano time is prone for overflows. Millis should be fine.

Reply:
could you please share more details (maybe some link to the description) regarding this issue?
it is a common pattern of nanotime usage..
Regarding overflow I am aware only about this point:
https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--
``Differences in successive calls that span greater than approximately 292 years (2^63 nanoseconds) will not correctly compute elapsed time due to numerical overflow.``
System.currentTimeMillis() may have not enough precision in some cases.